noir-projects/noir-protocol-circuits/crates/private-kernel-lib/src/private_kernel_reset.nr

```diff
@@ -1,97 +1,101 @@
     transient_note_hash_indexes_for_nullifiers: [u32; MAX_NULLIFIERS_PER_TX],
-    transient_note_hash_indexes_for_logs: [u32; MAX_NOTE_ENCRYPTED_LOGS_PER_TX],
     note_hash_read_request_hints: NoteHashReadRequestHints<NH_RR_PENDING, NH_RR_SETTLED>,
     nullifier_read_request_hints: NullifierReadRequestHints<NLL_RR_PENDING, NLL_RR_SETTLED>,
     key_validation_hints: [KeyValidationHint; KEY_VALIDATION_REQUESTS],
 }
 
 struct PrivateKernelResetCircuitPrivateInputs<NH_RR_PENDING, NH_RR_SETTLED, NLL_RR_PENDING, NLL_RR_SETTLED, KEY_VALIDATION_REQUESTS> {
     previous_kernel: PrivateKernelData,
-    outputs: PrivateKernelResetOutputs,
     hints: PrivateKernelResetHints<NH_RR_PENDING, NH_RR_SETTLED, NLL_RR_PENDING, NLL_RR_SETTLED, KEY_VALIDATION_REQUESTS>,
 }
 
 impl<NH_RR_PENDING, NH_RR_SETTLED, NLL_RR_PENDING, NLL_RR_SETTLED, KEY_VALIDATION_REQUESTS> PrivateKernelResetCircuitPrivateInputs<NH_RR_PENDING, NH_RR_SETTLED, NLL_RR_PENDING, NLL_RR_SETTLED, KEY_VALIDATION_REQUESTS> {
+    unconstrained fn generate_output(self) -> PrivateKernelResetOutputs {
+        ResetOutputComposer::new(
+            self.previous_kernel.public_inputs,
+            self.hints.transient_nullifier_indexes_for_note_hashes,
+            self.hints.transient_note_hash_indexes_for_nullifiers
+        ).finish()
+    }
+
     pub fn execute(self) -> PrivateKernelCircuitPublicInputs {
-        let mut previous_public_inputs = self.previous_kernel.public_inputs;
+        // Generate output.
+        let output = self.generate_output();
 
-        // verify/aggregate the previous kernel
-        self.previous_kernel.verify();
+        // Validate inputs.
+        if !std::runtime::is_unconstrained() {
+            // verify/aggregate the previous kernel
+            self.previous_kernel.verify();
+        }
 
-        previous_public_inputs.validation_requests = PrivateValidationRequestProcessor {
+        let previous_public_inputs = self.previous_kernel.public_inputs;
+        let updated_validation_requests = PrivateValidationRequestProcessor {
             validation_requests: previous_public_inputs.validation_requests,
             note_hash_read_request_hints: self.hints.note_hash_read_request_hints,
             pending_note_hashes: previous_public_inputs.end.note_hashes,
             note_hash_tree_root: previous_public_inputs.constants.historical_header.state.partial.note_hash_tree.root,
             nullifier_read_request_hints: self.hints.nullifier_read_request_hints,
             pending_nullifiers: previous_public_inputs.end.nullifiers,
             nullifier_tree_root: previous_public_inputs.constants.historical_header.state.partial.nullifier_tree.root,
-            key_validation_hints: self.hints.key_validation_hints,
+            key_validation_hints: self.hints.key_validation_hints
         }.validate();
 
-        verify_squashed_transient_note_hashes_and_nullifiers(
+        // Validate output.
```

Comment:
I think we can skip this if within unconstrained? To go a bit faster :partying_face:

Reply:
Good point! Will change 😄